cartoframes/core/cartodataframe.py

```diff
@@ -115,76 +118,88 @@ def to_carto(self, *args, **kwargs):
+        from ..viz import Map, Layer
+        return Map(Layer(self, *args, **kwargs))
 
-            Decode the geometry automatically:
+    def has_geometry(self):
+        return self._geometry_column_name in self
 
-            .. code::
-                from cartoframes import CartoDataFrame
+    def set_geometry(self, col, drop=False, inplace=False, crs=None):
+        if inplace:
+            frame = self
+        else:
+            frame = self.copy()
 
-                cdf = CartoDataFrame.from_file('filename.csv').convert()
+        # Decode geometry:
+        #   WKB, EWKB, WKB_HEX, EWKB_HEX, WKB_BHEX, EWKB_BHEX, WKT, EWKT
+        if isinstance(col, str) and col in frame:
+            frame[col] = decode_geometry_column(frame[col])
+        else:
+            col = decode_geometry_column(col)
 
-            Passing the geometry column explicitly:
+        # Call super set_geometry with decoded column
+        super(CartoDataFrame, frame).set_geometry(col, drop=drop, inplace=True, crs=crs)
 
-            .. code::
+        if not inplace:
+            return frame
 
-                from cartoframes import CartoDataFrame
+    def set_geometry_from_xy(self, x, y, drop=False, inplace=False, crs=None):
+        if isinstance(x, str) and x in self and isinstance(y, str) and y in self:
+            x_col = self[x]
+            y_col = self[y]
+        else:
+            x_col = x
+            y_col = y
 
-                cdf = CartoDataFrame.from_file('filename.csv')
-                cdf.convert(geom_column='my_geom_column')
+        # Generate geometry
+        geom_col = points_from_xy(x_col, y_col)
 
-            Passing lnglat_columns explicitly:
+        # Call super set_geometry with generated column
+        frame = super(CartoDataFrame, self).set_geometry(geom_col, inplace=inplace, crs=crs)
 
-            .. code::
+        if drop:
+            if frame is None:
+                frame = self
+            del frame[x]
+            del frame[y]
 
-                from cartoframes import CartoDataFrame
+        return frame
 
-                cdf = CartoDataFrame.from_file('filename.csv')
-                cdf.convert(lnglat_columns=['longitude', 'latitude'])
+    @property
+    def _constructor(self):
+        return CartoDataFrame
 
-            Passing the index column explicitly:
+    def __getitem__(self, key):
+        result = super(self._constructor, self).__getitem__(key)
+        if isinstance(key, (list, ndarray, Series)):
+            result.__class__ = self._constructor
+        return result
 
-            .. code::
+    def __finalize__(self, *args, **kwargs):
+        result = super(self._constructor, self).__finalize__(*args, **kwargs)
+        result.__class__ = self._constructor
+        return result
 
-                from cartoframes import CartoDataFrame
+    def astype(self, *args, **kwargs):
+        result = super(self._constructor, self).astype(*args, **kwargs)
+        result.__class__ = self._constructor
```

Comment:
Note that `astype` does not always return a GeoDataFrame on purpose (eg when there is no geometry column any more). You could also set the class only if `result` is actually a GeoDataFrame.

Reply:
Yes. This is a good point. More info here: https://github.com/CartoDB/cartoframes/issues/1260#issuecomment-559745122